capsules/src/mac.rs

```diff
@@ -0,0 +1,365 @@
+        buf[begin..begin + payload.len()].copy_from_slice(payload);
+        self.data_len += payload.len();
+
+        ReturnCode::SUCCESS
+    }
+}
+
+pub trait Mac {
+    fn get_address(&self) -> u16; //....... The local 16-bit address
+    fn get_address_long(&self) -> [u8; 8]; // 64-bit address
+    fn get_pan(&self) -> u16; //........... The 16-bit PAN ID
+    fn get_channel(&self) -> u8;
+    fn get_tx_power(&self) -> i8;
+
+    fn set_address(&self, addr: u16);
+    fn set_address_long(&self, addr: [u8; 8]);
+    fn set_pan(&self, id: u16);
+    fn set_channel(&self, chan: u8) -> ReturnCode;
+    fn set_tx_power(&self, power: i8) -> ReturnCode;
+
+    fn config_commit(&self) -> ReturnCode;
+
+    fn is_on(&self) -> bool;
+    fn prepare_data_frame(&self,
+                          buf: &mut [u8],
+                          dst_pan: PanID,
+                          dst_addr: MacAddress,
+                          src_pan: PanID,
+                          src_addr: MacAddress,
+                          security_needed: Option<(SecurityLevel, KeyId)>)
+                          -> Result<FrameInfo, ()>;
+    fn transmit(&self,
+                buf: &'static mut [u8],
+                frame_info: FrameInfo)
+                -> (ReturnCode, Option<&'static mut [u8]>);
+}
+
+pub trait TxClient {
+    fn send_done(&self, spi_buf: &'static mut [u8], acked: bool, result: ReturnCode);
+}
+
+pub trait RxClient {
+    fn receive<'a>(&self,
+                   buf: &'a [u8],
+                   header: Header<'a>,
+                   data_offset: usize,
+                   data_len: usize,
+                   result: ReturnCode);
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum TxState {
+    Idle,
+    Encrypting,
+    ReadyToTransmit,
+}
+
+pub struct MacDevice<'a, R: radio::Radio + 'a> {
+    radio: &'a R,
+    data_sequence: Cell<u8>,
+    config_in_progress: Cell<bool>,
+    tx_buf: TakeCell<'static, [u8]>,
+    tx_info: Cell<Option<FrameInfo>>,
+    tx_state: Cell<TxState>,
+    tx_client: Cell<Option<&'static TxClient>>,
+    rx_client: Cell<Option<&'static RxClient>>,
+}
+
+impl<'a, R: radio::Radio + 'a> MacDevice<'a, R> {
+    pub fn new(radio: &'a R) -> MacDevice<'a, R> {
+        MacDevice {
+            radio: radio,
+            data_sequence: Cell::new(0),
+            config_in_progress: Cell::new(false),
+            tx_buf: TakeCell::empty(),
+            tx_info: Cell::new(None),
+            tx_state: Cell::new(TxState::Idle),
+            tx_client: Cell::new(None),
+            rx_client: Cell::new(None),
+        }
+    }
+
+    pub fn set_transmit_client(&self, client: &'static TxClient) {
+        self.tx_client.set(Some(client));
+    }
+
+    pub fn set_receive_client(&self, client: &'static RxClient) {
+        self.rx_client.set(Some(client));
+    }
+
+    fn lookup_key(&self, level: SecurityLevel, key_id: KeyId) -> Option<(Security, [u8; 16])> {
```

Comment:
I believe this is still todo, right? i.e. it works for a certain static scenario but it's a kind of useless function as it is right now, that's basically here as a stub. If so, a comment would be nice saying that.